src/NHibernate/Hql/Ast/ANTLR/HqlSqlWalker.cs

```diff
@@ -1253,32 +1246,30 @@ public void Visit(IASTNode node)
 			//          2) here we would need to track each comparison individually, along with
 			//              the join alias to which it applies and then pass that information
 			//              back to the FromElement so it can pass it along to the JoinSequence
-			if ( node is DotNode ) 
+			if (_multiTable && node is DotNode dotNode)
 			{
-				DotNode dotNode = ( DotNode ) node;
 				FromElement fromElement = dotNode.FromElement;
-				if ( _referencedFromElement == null )
+				if (_joinFragment == fromElement)
 				{
-					_referencedFromElement = fromElement;
```

Comment:
Old logic sets `_referencedFromElement`  and `_joinAlias` for first occurrence of `FromElement` in with clause. That's strange and doesn't work correctly for reversed checks.
Works:
`left join root.Entity b with b.X = root.X`

Doesn't:
`left join root.Entity b with root.X = b.X`